### PR TITLE
Add simple prototype of functionality to push to spack caches

### PR DESCRIPTION
**Pull request by rfbgo**
I had a simple idea to push caches over breakfast, so threw together a quick prototype 

This is 90% glue but I wanted to experiment with how hard it would be to push a ramble generated env to a spack cache for later use (turns out its super easy!) 

This PR is more of a PoC and I can tidy it up a bit if we decide it's a feature we are interested in making mainline

### CODE REVIEW

File: lib/ramble/ramble/application_types/spack.py [hunk 1]
```diff
@@ -365,6 +365,21 @@ def _mirror_software(self, workspace):
             with self.logger.force_echo():
                 tty.die(e)
 
+    register_phase('push_to_spack_cache', pipeline='pushtocache', depends_on=[])
+
+    def _push_to_spack_cache(self, workspace):
+        try:
+            # TODO: make sure this is called as infrequently as possible
+            # TODO: this is a lot of repeated stuff, can we DRY with out pipelines?
```
Comment (douglasjacobsen) [inline lib/ramble/ramble/application_types/spack.py:373]: It is called a lot of times. Maybe there's a better way we can do it though.
I don't have any great suggestions on where this should go though. Maybe a wrapper within the runner?

You probably also want to make sure you deactivate the spack_runner.

File: lib/ramble/ramble/cmd/workspace.py [hunk 2]
```diff
@@ -51,6 +51,7 @@
     'concretize',
     'setup',
     'analyze',
+    'pushtocache',
```
Comment (douglasjacobsen) [inline lib/ramble/ramble/cmd/workspace.py:54]: Can names have hyphens?

File: lib/ramble/ramble/pipeline.py [hunk 3]
```diff
@@ -319,13 +319,33 @@ def _complete(self):
...
+
+    def _prepare(self):
+        super()._prepare()
+        self.workspace.spack_cache_path = self.spack_cache_path
+
+    def _complete(self):
+        tty.msg('Pushed envs to spack cache %s' % self.spack_cache_path)
+
+
+# TODO: should this use the classes `.name`?
```
Comment (douglasjacobsen) [inline lib/ramble/ramble/pipeline.py:340]: Yeah, I think there are several ways we can remove this. I think I only left it for error checking (i.e. trying to create a pipeline that doesn't exist yet)

File: lib/ramble/ramble/spack_runner.py [hunk 4]
```diff
@@ -676,6 +676,36 @@ def mirror_environment(self, mirror_path):
...
+    def push_to_spack_cache(self, spack_cache_path):
+        """Push packages for a given env to the spack cache"""
+        self._check_active()
+
+        hash_list = self.get_env_hash_list()
+
+        args = [
+            "buildcache",
+            "push",
+            "-a",  # TODO: remove this, it's just here for testing
```
Comment (douglasjacobsen) [inline lib/ramble/ramble/spack_runner.py:698]: I think we'll just need a way to pass arguments in to push. Maybe the same way we pass in args for install and concretize?

File: lib/ramble/ramble/application_types/spack.py (hunk 1 above)
Comment (rfbgo) [reply lib/ramble/ramble/application_types/spack.py:373]: Good catch on the deactivate.

I added an `if already_pushed` to the spack_runner which will help some

File: lib/ramble/ramble/cmd/workspace.py (hunk 2 above)
Comment (rfbgo) [reply lib/ramble/ramble/cmd/workspace.py:54]: Right now, in short, no 

The reason is that we directly map this dict to a function named `workspace_{key}` which would mean we need a function with hypens in. Not impossible to remap if we do want it but does not work out of the box today

File: lib/ramble/ramble/pipeline.py (hunk 3 above)
Comment (rfbgo) [reply lib/ramble/ramble/pipeline.py:340]: OK I will make a task to clean this up and remove the TODO here

File: lib/ramble/ramble/spack_runner.py (hunk 4 above)
Comment (rfbgo) [reply lib/ramble/ramble/spack_runner.py]: Good idea, done

File: lib/ramble/ramble/cmd/workspace.py (hunk 2 above)
Comment (rfbgo) [reply lib/ramble/ramble/cmd/workspace.py]: Updated this functionality in a seperate PR (#286) and added it here

**Review by douglasjacobsen — APPROVED** [review]

**Pull request merged**